elmerfem: make elmerfem-9.0 run properly

grindhold opened the pull request:
###### Description of changes

This is the result of my efforts to clear out the problems mentioned in #167599 . The import-problem was not the only one to fix in order to receive a working build of ElmerFEM:

  0) ElmerGUI displays things in the Model tree on the left screen side now [Before it was blank space which is not as intended by the developers]
  1) Enable the import of STEP- and other format by building against opencascade [This is used in many tutorials and will annoy users even if the software is theoretically usable without it]
  2) Patch the source to enable builds with gcc-11 and gfortran-11
  3) Build against Qt5  instead of Qt4 and supply relevant patches

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Wulfsta commented on pkgs/applications/science/physics/elmerfem/default.nix:
```diff
@@ -13,19 +13,42 @@ stdenv.mkDerivation rec {
+    qt5.wrapQtAppsHook
+  ];
+  buildInputs = [
+    mpi blas liblapack
```
Can you separate these onto their own lines?

Wulfsta commented on pkgs/applications/science/physics/elmerfem/default.nix:
```diff
@@ -13,19 +13,42 @@ stdenv.mkDerivation rec {
+    libsForQt5.qt5.qtbase
+    libsForQt5.qt5.qtscript
+    libsForQt5.qwt
+    libGL libGLU
```
See above.

Wulfsta reviewed (changes requested):
Looks good other than formatting!

grindhold commented:
i applied your one-dependency-per-line suggestion but i think i messed up the rebase. will fix tomorrowb

grindhold committed:
elmerfem: make elmerfem-9.0 properly

  0) ElmerGUI displays things in the Model tree on the
     left screen side again
  1) Enable the import of STEP- and other opencascade-
     based formats
  2) Patch the source to enable builds with gcc-11
  3) Build the Qt5 version instead of the Qt4 version

grindhold commented:
fixed the rebase

grindhold commented:
Is there anything left to do in order to get this merged?

Wulfsta commented:
Hmm, there’s a check that’s still running? Maybe try rerunning those?

grindhold commented:
i was told by a community member that the aarch64-darwin builldboxes were broken lately and have a huge backlog now.

Wulfsta commented:
It’s probably just going to take a little while to get picked up by someone with merge permissions then.

SuperSandro2000 commented on pkgs/applications/science/physics/elmerfem/default.nix:
```diff
@@ -13,19 +13,45 @@ stdenv.mkDerivation rec {
+    cmake
+    gfortran
+    pkg-config
+    git
```
Why do we need git?

Wulfsta replied:
We probably don’t, looks like an oversight from when I originally wrote this.

grindhold replied:
done

SuperSandro2000 commented on pkgs/applications/science/physics/elmerfem/default.nix:
```diff
@@ -13,19 +13,45 @@ stdenv.mkDerivation rec {
+  ];
 
   preConfigure = ''
     patchShebangs ./
```
Can we be more precise?

Wulfsta replied:
As I recall there are many files to patch, I don’t think this is an issue.

SuperSandro2000 replied:
> I don’t think this is an issue.

Well, it takes longer if the script looks at every file.

SuperSandro2000 commented on pkgs/applications/science/physics/elmerfem/default.nix:
```diff
@@ -13,19 +13,45 @@ stdenv.mkDerivation rec {
+    mpi
+    blas
+    liblapack
+    libsForQt5.qt5.qtbase
+    libsForQt5.qt5.qtscript
```
```suggestion
    libsForQt5.qtbase
    libsForQt5.qtscript
```

grindhold replied:
done

SuperSandro2000 commented on pkgs/applications/science/physics/elmerfem/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, cmake, git, gfortran, mpi, blas, liblapack, qt4, qwt6_qt4, pkg-config }:
+{ lib, stdenv, fetchFromGitHub, cmake, git, gfortran, mpi, blas, liblapack, qt5, pkg-config, libGL, libGLU, opencascade, libsForQt5, vtkWithQt5}:
```
Either use qt5 or libForQt5.

grindhold replied:
done
